### PR TITLE
Introduced `@ResourceLock(target = SELF | CHILDREN)`

### PR DESCRIPTION
**Pull request by anatolyD**
Defaults to SELF to preserve existing behavior.
Using CHILDREN has the same effect as declaring @ResourceLock on every @Test method of a test class.

Issue: #3102

## Overview

Introduced `@ResourceLock(target = SELF | CHILDREN)` where the target defaults to `SELF` to preserve existing behavior. Using `CHILDREN` has the same effect as declaring `@ResourceLock` on every `@Test` method of a test class.

I've added a couple of integration tests to verify it works, plus one for NodeWalker. Happy to hear the feedback. 😄 

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLockTarget.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package org.junit.jupiter.api.parallel;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import org.apiguardian.api.API;
+
+/**
+ *	Indicates the target of a {@link ResourceLock}.
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLockTarget.java:18]:
```suggestion
 * Indicates the target of a {@link ResourceLock}.
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLock.java [hunk 2]
```diff
@@ -40,6 +40,9 @@
  * <p>Since JUnit Jupiter 5.4, this annotation is {@linkplain Inherited inherited}
  * within class hierarchies.
  *
+ * <p>Since JUnit Jupiter 5.10, this annotation can be used to specify the target
+ * of the lock with {@link ResourceLockTarget}.
+ *
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLock.java:45]: What happens if a class has a `@ResourceLock` annotation with target `CHILDREN` and a method has one for the same resource? Does the one from the method override the one from the class? This should be documented here.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ExclusiveResource.java [hunk 3]
```diff
@@ -52,18 +52,34 @@ public class ExclusiveResource {
...
+	/**
+	 * Create a new {@code ExclusiveResource}.
+	 *
+	 * @param key the identifier of the resource; never {@code null} or blank
+	 * @param lockMode the lock mode to use to synchronize access to the
+	 * resource; never {@code null}
+	 * @param lockScope the lock scope to use to synchronize access to the
+	 * resource; never {@code null}
+	 */
+	public ExclusiveResource(String key, LockMode lockMode, LockScope lockScope) {
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ExclusiveResource.java:79]: I don't think this should be a Platform concept. Instead, Jupiter should return `ExclusiveResource` instances from the children `TestDescriptors` if their parent declares a `@ResourceLock` with `target` of `CHILDREN`.

**Review by marcphilipp — CHANGES_REQUESTED** [review]
Sorry for the delay! This looks very useful. 🙂

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ExclusiveResource.java (hunk 3 above)
Comment (anatolyD) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ExclusiveResource.java]: sounds good, no need to bother Platform with those details :) I drafted how it may look in https://github.com/junit-team/junit5/pull/3220/files#diff-63a01db8dd78b02e05c3888eb786c656af0cd3dc8263c816530003f2d7069daeR188

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java [hunk 4]
```diff
@@ -182,15 +181,32 @@ public static ExecutionMode toExecutionMode(org.junit.jupiter.api.parallel.Execu
...
-	Set<ExclusiveResource> getExclusiveResourcesFromAnnotation(AnnotatedElement element) {
+	protected List<ResourceLock> getResourceLocks() {
+		return Collections.emptyList();
+	}
+
+	protected Set<ExclusiveResource> collectExclusiveResourcesFromHierarchy() {
+		Set<ExclusiveResource> resources = mapResourceLocksForTarget(getResourceLocks(), ResourceLockTarget.SELF);
+
+		Optional<TestDescriptor> nextParent = getParent();
+		while (nextParent.isPresent()) {
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java:192]: IIUC `@ResourceLock(target = CHILDREN)` should only cause them to be applied to _direct_ children.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java (hunk 4 above)
Comment (Squibbles) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java]: > IIUC `@ResourceLock(target = CHILDREN)` should only cause them to be applied to _direct_ children.

Is there any reason for this?  Is it harmful in someway for it to not be just direct children?  e.g. in our project our tests inherit from base classes scattered throughout our packages and those base classes tend to inherit from a single source parent.  If I have a resource that I want read locked for everything (main server component for example) it makes more sense to me to have the lock specified in one place instead of having to remember to add it to all the base classes

Comment (marcphilipp) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java]: @Squibbles This is unrelated to inheritance in class hierarchies. The `@ResourceLock` annotation is inherited to subclasses already. However, the way resource locks are applied is that if a node acquires a lock, its children already hold it while being executed. Therefore it doesn't make sense to recursively declare it for all descendants.

**Review by marcphilipp — CHANGES_REQUESTED** [review]
I think we also need to think about the case where `@ResourceLock(target = CHILDREN)` is applied to methods. I think for `@Test` this should cause a failure. For test templates such as `@ParameterizedTest` and `@RepeatedTest` the invocations should get the `ExclusiveResource` (might already work with the code in this PR but we're missing a test). For `@TestFactory` methods, the direct `DynamicNodeTestDescriptor` children should get the `ExclusiveResource`.

**Comment by marcphilipp** [comment]
@anatolyD Do you have time to continue working on this PR or should someone from the JUnit team take over?

**Comment by anatolyD** [comment]
> @anatolyD Do you have time to continue working on this PR or should someone from the JUnit team take over?

Hey, yes I would love to continue, but if time is a pressuring matter you may want to take it over. I plan to get to it within a week or two

**Comment by marcphilipp** [comment]
I don't think it's not urgent so that's fine. 👍

**Comment by marcphilipp** [comment]
@anatolyD Are you still interested in continuing on this PR?

**Comment by anatolyD** [comment]
Hey, right, it should be closed. Will not progress on it

**Comment by marcphilipp** [comment]
No worries, thanks for letting us know!
